app/plugin_db_driver.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"sync"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+type DriverImpl struct {
+	s       *Server
+	connMut sync.RWMutex
+	connMap map[string]*sql.Conn
+	txMut   sync.Mutex
+	txMap   map[string]driver.Tx
+	stMut   sync.Mutex
+	stMap   map[string]driver.Stmt
+	rowsMut sync.Mutex
+	rowsMap map[string]driver.Rows
+}
+
+func NewDriverImpl(s *Server) *DriverImpl {
+	return &DriverImpl{
+		s:       s,
+		connMap: make(map[string]*sql.Conn),
+		txMap:   make(map[string]driver.Tx),
+		stMap:   make(map[string]driver.Stmt),
+		rowsMap: make(map[string]driver.Rows),
+	}
+}
+
+func (d *DriverImpl) Conn() (string, error) {
+	conn, err := d.s.sqlStore.GetMaster().Db.Conn(context.Background())
+	if err != nil {
+		return "", err
+	}
+	connID := model.NewId()
+	d.connMut.Lock()
+	d.connMap[connID] = conn
+	d.connMut.Unlock()
+	return connID, nil
+}
+
+func (d *DriverImpl) ConnPing(connID string) error {
+	d.connMut.RLock()
+	conn := d.connMap[connID]
```

Comment:
I see that we tend to assume the connection for the given id exists. Should we add a check instead and return error otherwise?

Reply:
Good point. A badly/evil plugin can knowingly crash the server.

Reply:
FWIW I think it could also happen as a simple result of concurrency (closing a connection while running something on the same).

Reply:
Indeed, good catch. 

> A Conn must call Close to return the connection to the database pool and may do so concurrently with a running query.